File: lib/dramatis/actor/behavior.rb
module Dramatis; end
module Dramatis::Actor; end

require 'dramatis/actor'

# The Dramatis::Actor::Behavior module is used to create actor behaviors.
# Creating an actor behavior does not create an actor but creates an object
# that an actor can "become" while still providing the actor method hook
# for value-add operations.

module Dramatis::Actor::Behavior

  def self.included cls #:nodoc:
    cls.instance_eval do
      include Dramatis
    end
    class << cls
      def new *args
        object = allocate
        eigenclass = ( class << object; self; end )
        eigenclass.send :define_method, :actor,
                ( lambda { Dramatis::Actor::Interface.new nil } )
        object.send :initialize, *args
        object
      end
    end
  end

  if false # for docs only ...

    # call-seq:
    #  actor -> an_interface
    #
    # actor provides classes that have mixed in Dramatis::Actor access
    # to a Dramatis::Actor::Interface object by which they can access
    # their actor name and other actor operations.

    def actor; end

  end

end

File: lib/dramatis/actor/behavior.py
from __future__ import absolute_import

from logging import warning

import dramatis
from dramatis.actor.name import Name as _Name
from dramatis import Runtime
import dramatis.runtime as runtime
from dramatis.actor.interface import Interface as _Interface

class Metaclass(type):
    
    def __init__(cls,name,bases,dict):
        super(Metaclass,cls).__init__(cls,name,bases, dict)
        
    def __call__( cls, *args, **kwds ):
        # See also actor comments ...
        interface = dramatis.Actor.Interface( None )
        behavior = cls.__new__( cls, *args, **kwds )
        class actor_class ( behavior.__class__ ):
            @property
            def actor( cls ):
                return interface
        behavior.__class__ = actor_class
        behavior.__init__( *args, **kwds )
        return behavior
        
class Behavior(object):
    """Class used as the base of actor behavior classes.

    Creating an actor behavior does not create an actor but creates an
    object that an actor can "become" while still providing the actor
    method hook or value-add operations."""

    __metaclass__ = Metaclass
